src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EnumerableExtensions.cs

```diff
@@ -3,40 +3,48 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
-using System.Linq;
+using System.Collections.Immutable;
 
 namespace System.Reflection.Internal
 {
+    /// <summary>
+    /// Replacements for System.Linq to avoid an unnecessary dependency. 
+    /// Parameter and return types strengthened to actual internal usage as an optimization.
+    /// </summary>
     internal static class EnumerableExtensions
     {
-        private class ComparisonComparer<T> : Comparer<T>
+        public static T FirstOrDefault<T>(this ImmutableArray<T> collection, Func<T, bool> predicate)
         {
-            private readonly Comparison<T> _compare;
-
-            public ComparisonComparer(Comparison<T> compare)
+            foreach (var item in collection)
             {
-                _compare = compare;
+                if (predicate(item))
+                {
+                    return item;
+                }
             }
 
-            public override int Compare(T x, T y)
-            {
-                return _compare(x, y);
-            }
+            return default(T);
         }
 
-        private static class Functions<T>
+        // used only in debugger display so we needn't get fancy with optimizations.
+        public static IEnumerable<TResult> Select<TSource, TResult>(this IEnumerable<TSource> source, Func<TSource, TResult> selector)
         {
-            public static readonly Func<T, T> Identity = t => t;
+            foreach (var item in source)
+            {
+                yield return selector(item);
+            }
         }
 
-        public static IOrderedEnumerable<T> OrderBy<T>(this IEnumerable<T> source, IComparer<T> comparer)
+        public static T Last<T>(this ImmutableArray<T>.Builder source)
         {
-            return source.OrderBy(Functions<T>.Identity, comparer);
+            return source[source.Count - 1];
         }
 
-        public static IOrderedEnumerable<T> OrderBy<T>(this IEnumerable<T> source, Comparison<T> compare)
+        public static List<T> OrderBy<T>(this List<T> source, Comparison<T> comparison)
         {
-            return source.OrderBy(new ComparisonComparer<T>(compare));
+            var list = new List<T>(source);
+            list.Sort(comparison);
+            return list;
         }
```

Comment:
See e.g. [https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Metadata/Ecma335/MetadataBuilder.Tables.cs,2070](https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Metadata/Ecma335/MetadataBuilder.Tables.cs,2070)

---
In reply to: [101688962](https://github.com/dotnet/corefx/pull/15366#discussion_r101688962) [](ancestors = 101688962)